pkg/controller/grouping/group_entity_index.go

```diff
@@ -0,0 +1,719 @@
+	AddExternalEntity(ee *v1alpha2.ExternalEntity)
+	// DeleteExternalEntity deletes an ExternalEntity from the index. If any existing groups are affected, eventHandlers
+	// will be called with the affected groups.
+	DeleteExternalEntity(ee *v1alpha2.ExternalEntity)
+	// AddNamespace adds or updates a Namespace to the index. If any existing groups are affected, eventHandlers will be
+	// called with the affected groups.
+	AddNamespace(namespace *v1.Namespace)
+	// DeleteNamespace deletes a Namespace to the index. If any existing groups are affected, eventHandlers will be
+	// called with the affected groups.
+	DeleteNamespace(namespace *v1.Namespace)
+	// Run starts the index.
+	Run(stopCh <-chan struct{})
+}
+
+type eventHandler func(group string)
+
+// entityType is an internal type used to differentiate Pod from ExternalEntity.
+type entityType string
+
+const (
+	podEntityType      entityType = "Pod"
+	externalEntityType entityType = "ExternalEntity"
+)
+
+// entityItem contains an entity (either Pod or ExternalEntity) and some relevant information.
+type entityItem struct {
+	// entity is either a Pod or an ExternalEntity.
+	entity metav1.Object
+	// labelKey is the key of the labelItem that the entityItem is associated with.
+	// entityItems will be associated with the same labelItem if they have same Namespace, entityType, and labels.
+	labelKey string
+}
+
+// labelItem represents an individual label set. It's the actual object that will be matched with label selectors.
+// Entities of same type in same Namespace having same labels will share a labelItem.
+type labelItem struct {
+	// The label set that will be used for matching.
+	labels labels.Set
+	// The namespace of the entities that share the labelItem.
+	namespace string
+	// The type of the entities that share the labelItem.
+	entityType entityType
+	// The keys of the entityItems that share the labelItem.
+	entities sets.String
+	// The keys of the selectorItems that match the labelItem.
+	selectors sets.String
+}
+
+// groupItem contains a group's metadata and its selector.
+type groupItem struct {
+	// The type of the group.
+	groupType string
+	// The name of the group. It must be unique within its own type.
+	name string
+	// The selector of the group.
+	selector *types.GroupSelector
+	// selectorKey is the key of the selectorItem that the groupItem is associated with.
+	// groupItems will be associated with the same selectorItem if they have same selector.
+	selectorKey string
+}
+
+// selectorItem represents an individual label selector. It's the actual object that will be matched with label sets.
+// Groups having same label selector will share a selectorItem.
+type selectorItem struct {
+	// The label selector that will be used for matching.
+	selector *types.GroupSelector
+	// The keys of the groupItems that share the selectorItem.
+	groups sets.String
+	// The keys of the labelItems that match the selectorItem.
+	labels sets.String
```

Comment:
could you rename this member to `labelKeys`?

Reply:
Sure, will make labelItem and its keys more explicit.